### ISSUE TITLE
Add more lower accuracy options

### ISSUE THREAD

**Issue by xoofx** [shibatch/sleef#227]
Hey,

This is a followup of issue [107](https://github.com/shibatch/sleef/issues/107#issuecomment-346933898)

It would be great to have more accuracy choice from instead of ULP10 and ULP35... The ULP10 is barely usable in a game, ULP35 is good, but is still sometimes too much:

- ULP10 is somewhat 23bits of accuracy  (for fp32)
- ULP35 is somewhat 19bits of accuracy

We are looking for accuracy below this (e.g 9 bits and even 4 bits).

How difficult would it be to add more accuracy choices to the library?

**Comment by shibatch** [comment]
Could you list specific functions to implement?

**Comment by xoofx** [comment]
> As for trig functions, reduction is taking time.
> Is it okay to have sinpi instead of sin?
> I need to know use cases.

Use cases is: When compiling a method, we can change the overall precision of mathematical function used transitively by an entry point function, so while having dedicated `sinpi` could be helpful in some cases (but we are not yet tracking range values of input/output in our compiler), the general case is that we should expect any values to come to a function like `sin` for example. Today we have ULP10 and ULP35 but we would like to go with lower precision. As a starter we could prioritize the following function (as mentioned in this [comment](https://github.com/shibatch/sleef/issues/187#issuecomment-397204189)):

> `cos`, `sin`, `log2`/`exp2` (`log`/`exp` but a bit less critical), `pow`, `acos`, `asin` (from higher to lower priority)

**Comment by shibatch** [comment]
What I meant is that sinpi/cospi would be a lot faster to compute than sin/cos.
I did a small experiment, and I think 20000 ULP sinpi and cospi are possible.

**Comment by shibatch** [comment]
And, I think 20000 ULP sincospif would be the best choice for trig functions.
sinpif and cospif would be not so much faster than sincospif.

**Comment by xoofx** [comment]
> What I meant is that sinpi/cospi would be a lot faster to compute than sin/cos.

Yep, but it requires changing user code which is a bit annoying. At some point we will do this switch transparently for the user (by detecting value ranges) but for now, we don't have an opportunity for that.
I'm fine with having the function very fast if the values is in the range of pi for example, and use a slightly slower route if not.

What I would like first is a general solution for standard values. As I pointed out, there is already a massive difference between ULP10 and ULP35 in terms of number of instructions, and this can be certainly improve when we ask for lower accuracy.

**Comment by xoofx** [comment]
> For sinf function, the polynomial for the kernel function only have 4 coefficients.
> I can reduce it to 3 or 2, but it would only make the computation faster by 4 or 8 clocks.

What I can see is that today for a function (I think it was for sin, but don't recall), we have these number of ASM instructions:
- u10: 385
- u35: 75

So if we go with much lower precision, I would expect this number to still go a bit down (specially if you go 9bits of accuracy)

**Comment by shibatch** [comment]
Okay, so I will check the argument is in the appropriate range, and then just multiply the arguments by PI.
Then, it would be much faster.

**Comment by shibatch** [comment]
I was testing this and it turned out that it is not so easy to make the trig functions faster.
One thing that would help is to stop measuring error in ULP, and set the absolute maximum error bound.
Does this work for you?

**Comment by xoofx** [comment]
> One thing that would help is to stop measuring error in ULP, and set the absolute maximum error bound.
Does this work for you?

Maybe, what would be the difference in practice compare to ULP?
ULP is nice to describe an error, but for a precision, I found easier to reason in terms of significand number of bits (so a ULP=1 for fp32 would be 23 bits)

**Comment by shibatch** [comment]
ULP is basically relative error.
What I am suggesting is absolute error.
The problem occurs, for example, when evaluating sin with argument close to n*PI/2.
In this case, the result will be a very small value, and thus the relative error becomes large.
If we specify the error bound in something like max(10ULP, 1e-8),
then it would be much easier to implement such a function.

**Comment by fpetrogalli-arm** [comment]
If high precision is not needed, how about providing FP16 version of the library? We would save computation, increase throughput, reduce energy... both Advanced Simd and SVE have these options, and it seems that x86 has some kind of support too.  FP16 is also very important for ML workloads.

As for changing the way we test SLEEF (no more ULPs)... I am slightly against it, and would like to find alternative solutions to support what this issue is requesting.

**Comment by shibatch** [comment]
FP16 is interesting, but my understanding is that that does not fit @xoofx 's need.
FP16C is not supported by sandybridge or prior processors.
There are a few similar but different formats, and this is also making the situation confusing.

What I am considering is to add new lower precision functions, and the new testing methods will be applied only to those new functions.

**Comment by xoofx** [comment]
> The problem occurs, for example, when evaluating sin with argument close to n*PI/2.
> In this case, the result will be a very small value, and thus the relative error becomes large.
> If we specify the error bound in something like max(10ULP, 1e-8),
> then it would be much easier to implement such a function.

Yeah, ok, that makes sense. Thanks. ULP is not really relevant in this case

**Comment by xoofx** [comment]
> If high precision is not needed, how about providing FP16 version of the library? We would save computation, increase throughput, reduce energy... both Advanced Simd and SVE have these options, and it seems that x86 has some kind of support too. FP16 is also very important for ML workloads.

I would not mind but this is also a bit orthogonal from lower precision

**Comment by shibatch** [comment]
So, I implemented fastsinf_u100000 and fastcosf_u100000.

https://github.com/shibatch/sleef/tree/Add_lowprec_sin_cos

Its error bound is max(10000 ULP, 1e-6).
The input domain is [-9, 9].
Finite math only.

My plan is also add fastsinf_u35 and fastcosf_u35 whose error bound is 1e-6 with full FP input domain.

**Comment by ax3l** [comment]
Hi,

I know this thread is a bit old, but for our scientific computations I would really benefit from ULP 35 variants (or lower precision) of atan, asinh, log for integrated Green's function solvers :) 
https://github.com/BLAST-WarpX/warpx/blob/26.06/Source/ablastr/fields/IntegratedGreenFunctionSolver.H#L36-L67
